File: server/api/bot.py
"""Bot configuration and statistics endpoints."""

from fastapi import APIRouter, Depends, HTTPException
from sqlalchemy import func, select
from sqlalchemy.ext.asyncio import AsyncSession

from server.db.connection import get_db
from server.db.models import Bot, BotStats
from server.security.auth import api_key_dependency
from server.security.input_validator import sanitize_bot_name, validate_color, validate_stats
from server.security.rate_limiter import rate_limit_api_calls, rate_limit_bot_config

from .schemas import BotConfigRequest, BotConfigResponse, BotStatsResponse

router = APIRouter(prefix="/api/v1/bot", tags=["bot"])


@router.put("/config", response_model=BotConfigResponse)
async def update_config(
    config: BotConfigRequest,
    bot: Bot = Depends(api_key_dependency),
    db: AsyncSession = Depends(get_db),
    _rl: None = Depends(rate_limit_bot_config),
) -> BotConfigResponse:
    """Update the authenticated bot's configuration."""
    if config.name is not None:
        bot.name = sanitize_bot_name(config.name)

    if config.avatar_color is not None:
        if not validate_color(config.avatar_color):
            raise HTTPException(status_code=422, detail="Invalid avatar color format.")
        bot.avatar_color = config.avatar_color

    if config.default_loadout is not None:
        loadout = config.default_loadout
        if not validate_stats(loadout.stats.model_dump()):
            raise HTTPException(
                status_code=422,
                detail="Invalid stats allocation. Total must be 20.",
            )
        bot.default_weapon = loadout.weapon.value
        bot.default_stats = loadout.stats.model_dump()
        bot.default_fallback = loadout.fallback_behavior.value

    await db.commit()
    await db.refresh(bot)

    return BotConfigResponse(
        name=bot.name,
        avatar_color=bot.avatar_color,
        default_weapon=bot.default_weapon,
        default_stats=bot.default_stats,
        default_fallback=bot.default_fallback,
        updated_at=bot.updated_at,
    )


@router.get("/stats", response_model=BotStatsResponse)
async def get_stats(
    bot: Bot = Depends(api_key_dependency),
    db: AsyncSession = Depends(get_db),
    _rl: None = Depends(rate_limit_api_calls),
) -> BotStatsResponse:
    """Return lifetime statistics for the authenticated bot."""
    stats = await db.get(BotStats, bot.id)
    if stats is None:
        raise HTTPException(status_code=404, detail="Stats not found for this bot.")

    kd_ratio = round(
        stats.kills / stats.deaths if stats.deaths > 0 else float(stats.kills),
        2,
    )

    rank_query = (
        select(func.count())
        .select_from(BotStats)
        .where(BotStats.elo > stats.elo)
    )
    result = await db.execute(rank_query)
    rank = result.scalar_one() + 1

    # Commit to persist last_seen update from auth dependency.
    await db.commit()

    return BotStatsResponse(
        kills=stats.kills,
        deaths=stats.deaths,
        kd_ratio=kd_ratio,
        assists=stats.assists,
        damage_dealt=stats.damage_dealt,
        damage_taken=stats.damage_taken,
        current_streak=stats.current_streak,
        best_streak=stats.best_streak,
        elo=stats.elo,
        rank=rank,
        time_alive_seconds=stats.time_alive_seconds,
        longest_life_secs=stats.longest_life_secs,
        rounds_played=stats.rounds_played,
        round_wins=stats.round_wins,
        pickups_collected=stats.pickups_collected,
        distance_traveled=stats.distance_traveled,
    )

File: go-arena/internal/api/bot.go
package api

import (
	"encoding/json"
	"io"
	"log/slog"
	"net/http"
	"time"

	"arena-server/internal/db"
	"arena-server/internal/game"
	"arena-server/internal/security"
)

// UpdateBotConfig handles PUT /api/v1/bot/config.
// It validates the incoming configuration and persists it to the database.
func UpdateBotConfig(w http.ResponseWriter, r *http.Request) {
	bot := security.GetBotFromContext(r.Context())
	if bot == nil {
		writeError(w, http.StatusUnauthorized, "authentication required")
		return
	}

	var req BotConfigRequest
	decoder := json.NewDecoder(r.Body)
	if err := decoder.Decode(&req); err != nil {
		writeError(w, http.StatusBadRequest, "invalid JSON body")
		return
	}
	if err := decoder.Decode(&struct{}{}); err != io.EOF {
		writeError(w, http.StatusBadRequest, "invalid JSON body")
		return
	}

	// Build and validate a complete candidate before changing the authenticated
	// bot record. A rejected field must not leave earlier fields applied.
	updatedBot := *bot

	// Apply name.
	if req.Name != nil {
		updatedBot.Name = security.SanitizeBotName(*req.Name)
	}

	// Apply avatar color.
	if req.AvatarColor != nil {
		if !security.ValidateColor(*req.AvatarColor) {
			writeError(w, http.StatusBadRequest, "invalid avatar_color: must be a hex color like #ff00aa")
			return
		}
		updatedBot.AvatarColor = *req.AvatarColor
	}

	// Apply loadout.
	if req.DefaultLoadout != nil {
		lo := req.DefaultLoadout

		if !security.ValidateWeapon(lo.Weapon) {
			writeError(w, http.StatusBadRequest, "invalid weapon")
			return
		}
		updatedBot.DefaultWeapon = lo.Weapon

		if err := security.ValidateStats(lo.Stats); err != nil {
			writeError(w, http.StatusBadRequest, "invalid stats: "+err.Error())
			return
		}
		updatedBot.DefaultStats = db.JSONBStats(lo.Stats)

		if !security.ValidateFallbackBehavior(lo.Fallback) {
			writeError(w, http.StatusBadRequest, "invalid fallback_behavior")
			return
		}
		updatedBot.DefaultFallback = lo.Fallback
	}

	updatedBot.UpdatedAt = time.Now()

	if db.Pool != nil {
		if err := db.UpdateBot(r.Context(), &updatedBot); err != nil {
			slog.Error("failed to update bot", "error", err, "bot_id", bot.ID)
			writeError(w, http.StatusInternalServerError, "failed to update bot")
			return
		}
	}
	*bot = updatedBot

	writeJSON(w, http.StatusOK, BotConfigResponse{
		BotID:       bot.ID,
		Name:        bot.Name,
		AvatarColor: bot.AvatarColor,
		Weapon:      bot.DefaultWeapon,
		Stats:       map[string]int(bot.DefaultStats),
		Fallback:    bot.DefaultFallback,
		UpdatedAt:   bot.UpdatedAt,
	})
}

// GetBotStats returns an http.HandlerFunc that serves GET /api/v1/bot/stats.
// It reads lifetime stats from the database and returns them alongside the
// bot's current rank.
func GetBotStats(engine *game.GameEngine) http.HandlerFunc {
	return func(w http.ResponseWriter, r *http.Request) {
		bot := security.GetBotFromContext(r.Context())
		if bot == nil {
			writeError(w, http.StatusUnauthorized, "authentication required")
			return
		}

		ctx := r.Context()

		// Default response when DB is unavailable.
		resp := BotStatsResponse{
			BotID: bot.ID,
			Name:  bot.Name,
			Elo:   1000,
		}

		if db.Pool != nil {
			stats, err := db.GetBotStats(ctx, bot.ID)
			if err != nil {
				slog.Error("failed to get bot stats", "error", err, "bot_id", bot.ID)
				writeError(w, http.StatusInternalServerError, "failed to get bot stats")
				return
			}

			if stats != nil {
				var kdRatio float64
				if stats.Deaths > 0 {
					kdRatio = float64(stats.Kills) / float64(stats.Deaths)
				} else {
					kdRatio = float64(stats.Kills)
				}

				rank, err := db.GetBotRank(ctx, bot.ID, "elo")
				if err != nil {
					slog.Warn("failed to get bot rank", "error", err, "bot_id", bot.ID)
				}

				resp = BotStatsResponse{
					BotID:            bot.ID,
					Name:             bot.Name,
					Kills:            stats.Kills,
					Deaths:           stats.Deaths,
					KDRatio:          kdRatio,
					Assists:          stats.Assists,
					DamageDealt:      stats.DamageDealt,
					DamageTaken:      stats.DamageTaken,
					CurrentStreak:    stats.CurrentStreak,
					BestStreak:       stats.BestStreak,
					Elo:              stats.Elo,
					Rank:             rank,
					RoundsPlayed:     stats.RoundsPlayed,
					RoundWins:        stats.RoundWins,
					PickupsCollected: stats.PickupsCollected,
					DistanceTraveled: stats.DistanceTraveled,
					TimeAliveSeconds: stats.TimeAliveSecs,
					LongestLifeSecs:  stats.LongestLifeSecs,
				}
			}
		}

		writeJSON(w, http.StatusOK, resp)
	}
}

// GetBotLive returns real-time game state for the authenticated bot.
// Returns whether the bot is connected, its current HP, position, action, etc.
func GetBotLive(engine *game.GameEngine) http.HandlerFunc {
	return func(w http.ResponseWriter, r *http.Request) {
		bot := security.GetBotFromContext(r.Context())
		if bot == nil {
			writeError(w, http.StatusUnauthorized, "authentication required")
			return
		}

		// Check if bot is currently in the game
		detail, ok := engine.GetBotDetail(bot.ID)
		if !ok {
			writeJSON(w, http.StatusOK, map[string]interface{}{
				"online":  false,
				"bot_id":  bot.ID,
				"name":    bot.Name,
				"message": "Bot is not currently connected to the arena",
			})
			return
		}

		// Get arena status for context
		gs := engine.GetFullGameState()
		phase, _ := gs["round_phase"].(string)

		// Build radar metrics (0-100 scale for visualization)
		roundKills, _ := detail["round_kills"].(int)
		roundDeaths, _ := detail["round_deaths"].(int)
		dmgDealt, _ := detail["round_damage_dealt"].(float64)
		dmgTaken, _ := detail["round_damage_taken"].(float64)
		shotsFired, _ := detail["round_shots_fired"].(int)
		shotsHit, _ := detail["round_shots_hit"].(int)
		distance, _ := detail["round_distance"].(float64)
		pickups, _ := detail["round_pickups"].(int)
		hp, _ := detail["hp"].(float64)
		maxHP, _ := detail["max_hp"].(float64)
		speed, _ := detail["speed"].(float64)
		atkMult, _ := detail["attack_multiplier"].(float64)
		defRed, _ := detail["defense_reduction"].(float64)
		killStreak, _ := detail["kill_streak"].(int)

		var accuracy float64
		if shotsFired > 0 {
			accuracy = float64(shotsHit) / float64(shotsFired) * 100
		}
		var dmgRatio float64
		if dmgTaken > 0 {
			dmgRatio = dmgDealt / dmgTaken
		} else if dmgDealt > 0 {
			dmgRatio = 10
		}

		// Action distribution from history
		actionCounts := detail["action_counts"]

		writeJSON(w, http.StatusOK, map[string]interface{}{
			"online":             true,
			"bot_id":             bot.ID,
			"name":               bot.Name,
			"phase":              phase,
			"hp":                 hp,
			"max_hp":             maxHP,
			"position":           detail["position"],
			"weapon":             detail["weapon"],
			"is_alive":           detail["is_alive"],
			"speed":              speed,
			"attack_mult":        atkMult,
			"defense_red":        defRed,
			"kill_streak":        killStreak,
			"round_kills":        roundKills,
			"round_deaths":       roundDeaths,
			"round_damage_dealt": dmgDealt,
			"round_damage_taken": dmgTaken,
			"round_shots_fired":  shotsFired,
			"round_shots_hit":    shotsHit,
			"round_distance":     distance,
			"round_pickups":      pickups,
			"accuracy":           accuracy,
			"damage_ratio":       dmgRatio,
			"active_effects":     detail["active_effects"],
			"dodge_cooldown":     detail["dodge_cooldown"],
			"cooldown_remaining": detail["cooldown_remaining"],
			"current_action":     detail["current_action"],
			"last_action_result": detail["last_action_result"],
			"frozen":             detail["frozen"],
			"action_counts":      actionCounts,
		})
	}
}
